### PR TITLE
[DI] Do not process bindings in AbstractRecursivePass

### PR DESCRIPTION
**Pull request by chalasr**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #24559 
| License       | MIT
| Doc PR        | n/a

### CODE REVIEW

**Comment by chalasr** [comment]
Build failure unrelated.

**Comment by dmaicher** [comment]
Does this mean that in the example mentioned in #24559 the inlined `bar` service is not removed anymore within `RemoveUnusedDefinitionsPass` with this fix?

**Comment by chalasr** [comment]
@dmaicher yes

**Comment by dmaicher** [comment]
:+1: Should it not be merged into 3.4 instead of master?

**Review by dmaicher — APPROVED** [review]

**Comment by stof** [comment]
Are you sure about it ? Bindings are used by the autowiring layer, which runs *before* the removal passes. If the autowiring has not used the binding to set a reference, we don't have a usage (and we don't need to prevent removal AFAIK). And if the binding was used to fill an argument, we will have a reference in the arguments (and so processed by the existing code)

**Comment by dmaicher** [comment]
> And if the binding was used to fill an argument, we will have a reference in the arguments (and so processed by the existing code)

@stof if the service was inlined this it not true I believe. It will have an inlined `Definition` in the arguments and not a `Reference`.

**Comment by stof** [comment]
@dmaicher and this is fine as well, as the argument is still filled properly. It does not need the binding anymore.

Actually, I think the issue is that AbstractRecursivePass is always processing bindings in the exact same way it processes arguments, method calls and so on.
And so, it will treat any reference in a binding as a usage of the service. But once bindings are used to fill arguments, they become useless. So they should not be visited by the validator passes checking for circular references, invalid references and so on.
I'm not even sure there is a single case where a recursive pass should visit them.

**Comment by dmaicher** [comment]
@stof ok I see :wink: Then we should indeed not check the References contained in bindings I guess. So the Definitions can be removed correctly for inlined services.

**Comment by chalasr** [comment]
PR updated to not process bindings in AbstractRecursivePass.

File: src/Symfony/Component/DependencyInjection/Compiler/AbstractRecursivePass.php [hunk 1]
```diff
@@ -67,7 +67,6 @@ protected function processValue($value, $isRoot = false)
             $value->setArguments($this->processValue($value->getArguments()));
             $value->setProperties($this->processValue($value->getProperties()));
             $value->setMethodCalls($this->processValue($value->getMethodCalls()));
-            $value->setBindings($this->processValue($value->getBindings()));
```
Comment (dmaicher) [inline src/Symfony/Component/DependencyInjection/Compiler/AbstractRecursivePass.php]: can we have a test case for this? for example within the `RemoveUnusedDefinitionsPassTest`?

Comment (dmaicher) [reply src/Symfony/Component/DependencyInjection/Compiler/AbstractRecursivePass.php]: Sorry I mean the pass that checks if services exist ;p

Comment (chalasr) [reply src/Symfony/Component/DependencyInjection/Compiler/AbstractRecursivePass.php]: Test added

File: src/Symfony/Component/DependencyInjection/Tests/Compiler/CheckExceptionOnInvalidReferenceBehaviorPassTest.php [hunk 2]
```diff
@@ -88,6 +89,22 @@ public function testProcessThrowsExceptionOnNonSharedUninitializedReference()
         $this->process($container);
     }
 
+    public function testProcessDefinitionWithBindings()
+    {
+        $container = new ContainerBuilder();
+
+        $container->register('a', 'stdClass');
```
Comment (dmaicher) [inline src/Symfony/Component/DependencyInjection/Tests/Compiler/CheckExceptionOnInvalidReferenceBehaviorPassTest.php]: this test will now also pass without your change I believe? we should remove this so the Reference `a` in the bindings does actually not exist on the container?

Comment (chalasr) [reply src/Symfony/Component/DependencyInjection/Tests/Compiler/CheckExceptionOnInvalidReferenceBehaviorPassTest.php]: oops :) fixed

**Review by dmaicher — APPROVED** [review]

**Review by nicolas-grekas — APPROVED** [review]

**Review by ogizanagi — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @chalasr.

**Pull request merged**